device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_025: [The `sendEvent` method shall create and attach the d2c link if necessary.]*/
+            if (!this._d2cLink) {
+              this._amqp.attachSenderLink(this._d2cEndpoint, null, (err, link) => {
+                if (err) {
+                  handleResult('AMQP Transport: Could not send', sendCallback)(err);
+                } else {
+                  debug('got a new D2C link');
+                  this._d2cLink = link;
+                  this._d2cLink.on('error', this._d2cErrorListener);
+                  this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
+                }
+              });
+            } else {
+              debug('using existing d2c link');
+              this._d2cLink.send(amqpMessage, handleResult('AMQP Transport: Could not send', sendCallback));
+            }
+          },
+          updateSharedAccessSignature: (sharedAccessSignature, updateSasCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_06_010: [If the AMQP connection is established, the `updateSharedAccessSignature` method shall call the amqp transport `putToken` method with the first parameter `audience`, created from the `sr` of the shared access signature, the actual shared access signature, and a callback.]*/
+            this._amqp.putToken(SharedAccessSignature.parse(sharedAccessSignature, ['sr', 'sig', 'se']).sr, sharedAccessSignature, (err) => {
+              if (err) {
+                this._amqp.disconnect(() => {
+                  updateSasCallback(getTranslatedError(err, 'AMQP Transport: Could not authorize with puttoken'));
+                });
+              } else {
+                /*Codes_SRS_NODE_DEVICE_AMQP_06_011: [The `updateSharedAccessSignature` method shall call the `done` callback with a null error object and a SharedAccessSignatureUpdated object as a result, indicating the client does NOT need to reestablish the transport connection.]*/
+                updateSasCallback(null, new results.SharedAccessSignatureUpdated(false));
+              }
+            });
+          },
+          getTwinReceiver: (callback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_026: [** The `getTwinReceiver` method shall call the `done` callback with a `null` error argument and the `AmqpTwinClient` instance currently in use.]*/
+            callback(null, this._twinClient);
+          },
+          sendTwinRequest: (method, resource, properties, body, sendTwinRequestCallback) => {
+            this._twinClient.sendTwinRequest(method, resource, properties, body, sendTwinRequestCallback);
+          },
+          attachC2DLink: (callback) => {
+            debug('attaching C2D link');
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_030: [The `Amqp` object shall attach the C2D `ReceiverLink` object if necessary to start receiving messages.]*/
+            this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, receiverLink) => {
```

Comment:
>attachReceiverLink [](start = 23, length = 18)

same comment about concurrency.  attachingReceiverLink feels like it should be a state.

Reply:
same answer.

---
In reply to: [141131946](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141131946) [](ancestors = 141131946)